### ISSUE TITLE
Output layer loss function gradient calculation (MCXENT)

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#349]
Calculation of output layer loss function derivatives appear to be incorrect for MCXENT. Other cases may be incorrect also, but I haven't checked the mathematics.
Basically, I'm of the opinion that the weight gradient (line 136) should be ```pYGivenX.sub(labels);``` not ```labels.sub(pYGivenX);```, similarly for bias calculation, OutputLayer.gradient() line 106 (```INDArray dy = labels.sub(netOut);```). Though this bias calculation is shared by all loss functions (hence I'm not sure if changing it for MCXENT would make others wrong or not).

Partial derivation (both for my own benefit and ease of communication) follows.
Denote:
![](http://mathurl.com/q8nt7yw.png) as weight from unit i (prev. layer) to unit j (output layer).
![](http://mathurl.com/nr9pktp.png) is input to unit j
![](http://mathurl.com/osglszn.png) is target/label
![eqn](http://mathurl.com/oopsolc.png) is output (i.e., softmax probability for class j in this case)
L loss function (standard cross-entropy loss function)

For a single example,
![eqn](http://mathurl.com/p3bd47v.png)
(and similar for bias)

![eqn](http://mathurl.com/na2dhq3.png), ![eqn](http://mathurl.com/o9kfnu5.png)

Now, ![eqn](http://mathurl.com/oh4uq3m.png) is non-trivial to derive. But according to a number of sources, we end up with (output - target) for the cross-entropy/softmax case:
![eqn](http://mathurl.com/o4wmc8o.png)
Refer to (though careful with the differences in notation):
[1] Hinton's NN course: https://youtu.be/UI-QnjgMgNY?t=6m27s
[2] http://neuralnetworksanddeeplearning.com/chap3.html, eq. (84)
[3] Bishop (1995) Neural Networks for Pattern Recognition (pg240, eq. 6.163) (Can't find an online version of this)

Thus we have:
![eqn](http://mathurl.com/octgupm.png)

**Comment by AlexDBlack** [comment]
As for your reference in the comment on my previous issue http://deeplearning.stanford.edu/wiki/index.php/Logistic_Regression_Vectorization_Example
Consider the stated weight update rule (i.e., add an amount proportional to gradient) vs. the neural network parameter update rule (subtract an amount proportional to gradient). So they are in fact equivalent, with (output-target) what we want for neural networks.

**Comment by agibsonccc** [comment]
https://github.com/deeplearning4j/deeplearning4j/commit/bfd4632e58feaed58cdbc9e1ec8087082bf3bfc1
